File: cis_interface/examples/rpcFib/src/rpcFibSrv.c

#include "PsiInterface.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>


int main(int argc, char *argv[]) {

  float timeSleep = atof(argv[1]);
  printf("Hello from C rpcFibSrv: sleeptime = %f\n", timeSleep);
  
  // Create server-side rpc conneciton using model name
  psiRpc_t rpc = psiRpcServer("rpcFibSrv", "%d", "%d %d");

  // Continue receiving requests until error occurs (the connection is closed
  // by all clients that have connected).
  int input;
  while (1) {
    int ret = rpcRecv(rpc, &input);
    if (ret < 0) {
      printf("rpcFibSrv(C): end of input\n");
      break;
    }

    // Compute fibonacci number
    printf("rpcFibSrv(C): <- input %d", input);
    int result = 1;
    int prevResult = 1;
    int prevPrev = 0;
    int idx = 1;
    while(idx++ < input){
      result = prevResult + prevPrev;
      prevPrev = prevResult;
      prevResult = result;
    }
    printf(" ::: ->(%2d %2d)\n", input, result);

    // Sleep and then send response back
    if (timeSleep) 
      sleep(timeSleep);
    rpcSend(rpc, input, result);
  }

  printf("Goodbye from C rpcFibSrv\n");
  
}

File: yggdrasil/examples/rpcFib/src/rpcFibSrv.f90
program main
  use fygg

  type(yggcomm) :: rpc
  character(len=32) :: arg
  real :: time_sleep
  integer :: input
  integer :: result0, prevResult, prevPrev, idx
  logical :: ret

  call get_command_argument(1, arg)
  read(arg, *) time_sleep
  write(*, '("Hello from Fortran rpcFibSrv: sleeptime = ",f10.4)') time_sleep

  ! Create server-side rpc conneciton using model name
  rpc = ygg_rpc_server("rpcFibSrv", "%d", "%d %d")

  ! Continue receiving requests until error occurs (the connection is
  ! closed by all clients that have connected).
  do while (.true.)
     write(*, '("rpcFibSrv(F): receiving...")')

     ret = ygg_recv_var(rpc, yggarg(input))
     if (.not.ret) then
        write(*, '("rpcFibSrv(F): end of input")')
        exit
     end if

     ! Compute fibonacci number
     result0 = 1
     prevResult = 1
     prevPrev = 0
     idx = 1
     do while (idx.lt.input)
        result0 = prevResult + prevPrev
        prevPrev = prevResult
        prevResult = result0
        idx = idx + 1
     end do
     write(*, '("rpcFibSrv(F): <- input ",i2," ::: ->(",i2," ",i2,")")') &
          input, input, result0

     ! Sleep and then send response back
     if (time_sleep.gt.0) call fsleep(int(time_sleep))
     ret = ygg_send_var(rpc, [yggarg(input), yggarg(result0)])
     if (.not.ret) then
        write(*, '("rpcFibSrv(F): ERROR sending")')
        exit
     end if

  end do

  write(*, '("Goodbye from Fortran rpcFibSrv")')

end program main
